File: Programs_in_Cpp/bellman_ford.cpp
// Program to implement Bellman Ford algorithm
#include <iostream>
#include <vector>
#include <limits>

using namespace std;

typedef struct Edge {
    int source;
    int destination;
    int weight;
} Edge;

class Graph {
public:
    int vertexNumber, edgeNumber;
    vector<Edge> graph;
    vector<int> distance;

    Graph(int m, int n) {
        vertexNumber = m;
        edgeNumber = n;
        distance.push_back(0);
        for (int i = 1; i < m; i++) {
            distance.push_back(INT_MAX);
        }
    }

    void printDistance() {
        cout << "Vertex\t\tDistance from source" << endl;
        for (int i = 0; i < vertexNumber; i++) {
            cout << i << "\t\t" << distance.at(i) << endl;
        }
    }

    void bellmanFord(int src) {
        for (int i = 1; i <= vertexNumber - 1; i++) {
            for (int j = 0; j < edgeNumber; j++) {
                int u = graph.at(j).source;
                int v = graph.at(j).destination;
                int w = graph.at(j).weight;
                if (distance.at(u) != INT_MAX && distance.at(u) + w < distance.at(v)) {
                    distance.at(v) = distance.at(u) + w;
                }
            }
        }

        for (int i = 0; i < edgeNumber; i++) {
            int u = graph.at(i).source;
            int v = graph.at(i).destination;
            int w = graph.at(i).weight;
            if (distance.at(u) != INT_MAX && distance.at(u) + w < distance.at(v)) {
                cout << "Graph contains negative weight cycle" << endl;
                return;
            }
        }

        printDistance();
    }
};

int main() {
    int V, E;
    cout << "Enter the number of vertices and edges: " << endl;
    cin >> V >> E;
    Edge edg;
    Graph g = Graph(V, E);
    cout << "Enter the source, destination and weight of the edge in a single line: " << endl;
    for(int i=0; i<E; i++) {
        cin >> edg.source >> edg.destination >> edg.weight;
        g.graph.push_back(edg);
    }
    int src;
    cout << "Enter the source vertex:" << endl;
    cin >> src;
    g.bellmanFord(src);
    return 0;
}

File: Part_A/bellman_ford.py
# Program to implement Bellman Ford algorithm
from sys import maxsize


# Class denoting the graph for the algorithm to be applied
class Graph:
    def __init__(self, vertices):
        self.V = vertices  # Initializing the number of vertices
        self.graph = []

    def add_edge(self, src, destination, wt):
        self.graph.append([src, destination, wt])  # Adding an edge to the graph list

    def print_distance(self, dist):
        print("Vertex\t\tDistance from source")
        for i in range(self.V):
            print("{}\t\t{}".format(i, dist[i]))  # Printing the result

    def bellman_ford(self, src):
        dist = [maxsize] * self.V
        dist[src] = 0

        for i in range(self.V - 1):
            for u, v, w in self.graph:
                if dist[u] != maxsize and dist[u] + w < dist[v]:
                    dist[v] = dist[u] + w

        for u, v, w in self.graph:
            if dist[u] != maxsize and dist[u] + w < dist[v]:
                print("Graph contains negative weight cycle")  # Checking if negetive weighted cycle is present
                return

        self.print_distance(dist)


if __name__ == "__main__":
    vertex_no = int(input("Enter the number of vertices of the graph:"))
    edges = int(input("Enter the number of edges of the graph:"))
    input_graph = Graph(vertex_no)
    print("Enter the edge in the format SOURCE, DESTINATION and WEIGHT in a SINGLE LINE:")
    for _ in range(edges):
        source, des, weight = map(int, input().split())
        input_graph.add_edge(source, des, weight)

    source = int(input("Enter the source vertex to begin the algorithm:"))
    input_graph.bellman_ford(source)